### DISCUSSION TITLE
Adding new oauth providers

### DISCUSSION THREAD

**Discussion by amend07** [apache/cloudstack#9129] [Ideas]
Hello there,

I just want to add keycloak as identity provider in oauth2 configuration. I add it in the **oauth.plugins** in the web interface. when i set up it in **OAuth provider** configuration under **Provider** section I can't find it. Is there any configuration that I miss to configure ? please reach me ASAP.

With regards,

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Reply by weizhouapache** [reply]
@amend07 
currently only google and github are supported.
you can add a provider similar as 
https://github.com/apache/cloudstack/tree/main/plugins/user-authenticators/oauth2/src/main/java/org/apache/cloudstack/oauth2/google
https://github.com/apache/cloudstack/tree/main/plugins/user-authenticators/oauth2/src/main/java/org/apache/cloudstack/oauth2/github

**Comment by shwstppr** [comment]
@amend07 not sure but currently only Google an GitHub are supported as plugins. @harikrishna-patnala is it possible to add keycloak?

**Comment by harikrishna-patnala** [comment]
@amend07 currently keycloak is not supported (only google and github), in order to have that support, it is needed to implement the keycloak plugin under "user-authenticators" > "oauth2".

**Comment by amend07** [comment]
> @amend07 currently keycloak is not supported (only google and github), in order to have that support, it is needed to implement the keycloak plugin under "user-authenticators" > "oauth2".

Thank you @harikrishna-patnala. So can I implement the keycloak plugin or should I have to wait for you to update the plugin?

**Comment by harikrishna-patnala** [comment]
@amend07 you are welcome to implement the plugin and raise a PR, I would say it is quite easy to implement I can surely help you review. All you need to do is implement "UserOAuth2Authenticator" interface and add couple of UI changes. Let me know your thoughts.

**Comment by amend07** [comment]
@harikrishna-patnala am happy to work with you. help me to join the team. thanks again for your support

**Comment by shwstppr** [comment]
@amend07 can we close this or maybe convert it to a discussion?

**Comment by amend07** [comment]
@shwstppr I am happy. create a discussion and mention me.

**Reply by alwayshusslin** [reply]
did this ever get implemented? we are looking to do the same

**Reply by DaanHoogland** [reply]
@alwayshusslin no issue or pr was created. maybe someone is working on it 🤷

**Comment by tazouxme** [comment]
That's an old topic but I'd be glad to provide my contribution and implement it.
What's the best way to contribute ? Starting with an issue or something else ?

**Reply by DaanHoogland** [reply]
@tazouxme , you can create an issue and enter a design description or go ahead and create a PR with a draft. It only depends on your prefered way of developing. Just be sure to ask for plenty of feedback early, to prevent having to do too much rework later.
